### DISCUSSION TITLE
Integrating with Google/OSS-fuzz

### DISCUSSION THREAD

**Discussion by jesslatimer** [bokeh/bokeh#12819] [General]
Hi,
I would like to help keep bokeh running smoothly by adding fuzz testing and integrating it with google/oss-fuzz. In case you are not familiar with fuzz testing, briefly, it is an additional layer of testing (in addition to unit testing) that will help uncover any hidden or hard-to-reach bugs that might have been missed during unit testing. [Wikipedia ](https://en.wikipedia.org/wiki/Fuzzing) explains it nicely:

> In programming and software development, fuzzing or fuzz testing is an automated software testing technique that involves providing invalid, unexpected, or random data as inputs to a computer program. The program is then monitored for exceptions such as crashes, failing built-in code assertions, or potential memory leaks.

Google offers a free, continuous fuzzing service called OSS-fuzz. If bokeh is integrated into oss-fuzz, the fuzz tests under bokeh will be built and then run once a day, to search for bugs and vulnerabilities in bokeh. This service can be integrated with the CI for bokeh, so that the fuzz tests are run for 10min or so for every pull request, preventing buggy code from being merged.

I've opened up a pull request to add a basic fuzz-testing harness here https://github.com/bokeh/bokeh/pull/12811. If you are keen on adding bokeh to oss-fuzz I'd be happy to champion the integration :)

**Comment by bryevdv** [comment]
cc @bokeh/dev for discussion

At least for me personally, I would want to see a concrete, targeted, and scoped proposal. Bokeh has nearly 10k unit tests, just on the Python side. Re-instrumenting all of them would would be an insurmountable task. But just randomly adding a single test or two here and there also does not seem worth the overhead and complication of dealing with yet another service integration. Before embarking on something like this I'd like to see clearly enunciated strategy that guides:

* what kinds of codepaths in bokeh would receive the most value from the coverage 
* where it should be applied to retroactively
* what cases it should to be applied in the future

laid out and documented in the dev guide.

**Comment by jesslatimer** [comment]
Not a problem. It sounds like you are after some specific information in this proposal, could you please send an example proposal?

**Reply by bryevdv** [reply]
@jesslatimer I'm not sure why you are asking for, the information I am interested in is what I listed in the response above. I want a plan, of some sort. The current PR adds a single random test file, in a random location completely outside our current repo test structure. At a minimum it needs to be decided where these tests go: are they a whole new separate class of tests (so a new dir `tests/fuzz`)? Are they augmentations of some subset of existing unit tests (so maybe a pytest marker for tests under `tests/unit`)? 

Either way, given limited resources, what are the most important areas of testing and API surface to focus us? Where is the most benefit to be had, in some detail? Is there a policy we should adopt going forward about certain kinds of codepaths that should encourage or require fuzz testing, specifically? I am not an expert about fuzz testing&mdash;this is something you will have to bring to the table.

**Comment by ianthomas23** [comment]
I'd be happy for any new testing to be run on the Bokeh codebase outside of our CI, and any problems found to be reported as github issues in the usual way. But I would not want to add it to our CI as it is already complicated enough. It might be fine if an existing or new maintainer were to commit to taking full responsibility for maintaining it going forward, but that seems unlikely.

Personally I am not a fan of nondeterministic tests as part of CI workflow. Bokeh is already difficult enough for newcomers to contribute to without the danger that their first PR fails CI with an unrelated failure.

**Reply by mattpap** [reply]
In CI we could have additional workflows behind `workflow_dispatch`, either triggered through GitHub's UI (like we do for release builds) or (possibly) via a special comment in a PR (that would require some configuration). I've been investigating this for reducing the default number of jobs, especially MacOS jobs, that we run for each PR and each merge.

**Reply by bryevdv** [reply]
`workflow_dispatch` seems like a reasonable way to keep things at arms length and not trigger on every push, which I agree should be avoided, certainly to start.

**Comment by bryevdv** [comment]
It seems unlikely that there will be any movement here so I will close this discussion
